src/defs.rs

```diff
@@ -53,6 +57,76 @@ pub struct WorkItem {
     pub name: String,
 }
 
+#[derive(Debug)]
+pub enum ArchiveType {
+    Zip(ZipArchive<File>),
+    Dir(PathBuf),
+}
+
+#[derive(Debug)]
+pub struct Archive {
+    pub item: ArchiveType,
+}
+
+impl Archive {
+
+    pub fn read_in_buffer(&mut self, name: &str, buf: &mut Vec<u8>) -> bool {
+        match self.item {
+            ArchiveType::Zip(ref mut zip) => {
+                match zip.by_name(name) {
+                    Ok(mut f) => {
+                        f.read_to_end(buf).expect("Failed to read gcda file");
+                        true
+                    },
+                    Err(_) => false,
+                }
+            },
+            ArchiveType::Dir(ref dir) => {
+                match File::open(dir.join(PathBuf::from(name))) {
+                    Ok(mut f) => {
+                        f.read_to_end(buf).expect("Failed to read gcda file");
+                        true
+                    },
+                    Err(_) => false,
+                }
+            }
+        }
+    }
+
+    pub fn extract(&mut self, name: &str, path: &PathBuf) -> bool {
+        match self.item {
+            ArchiveType::Zip(ref mut zip) => {
+                match zip.by_name(name) {
+                    Ok(mut f) => {
+                        let mut file = File::create(&path).expect("Failed to create file");
+                        io::copy(&mut f, &mut file).expect("Failed to copy file from ZIP");
+                        true
+                    },
+                    Err(_) => false,
+                }
+            },
+            ArchiveType::Dir(ref dir) => {
+                let src_path = dir.join(PathBuf::from(name));
+                let dest_parent = path.parent().unwrap();
+                if !dest_parent.exists() {
+                    match fs::create_dir_all(dest_parent) {
```

Comment:
It should be faster if we try to create the directory without checking if it exists first, and then ignore the error if it already exists.